README.md

# ABOS (AVR Bootloader Open Source)

### Simple bootloader to avoid ISP hardware to program AVR Microcontrollers


## Supported Devices
- ATmega16
- ATmega32
- ATmega128/ATmega128A


### The default baudrate is **_38400_**, is defined in the firs code line of the **_"abos.c"_** file. If you need other baudrate, change these value.
---

# Installation

1. Download the repository:

        git clone https://github.com/alfreedom/abos

2. Edit the include device configuration in the Makefile to choose the microcontroller:
        
        include devices/myDevice.mk

3. Change the **AVRDUDE_PROG** to specify your programmer in the Makefile:

        AVRDUDE_PROG = -c usbtiny

4. Enter to the bootloader folder and run:

        make all
        make install

        or

        make
        make flash
        make fuse
        make lockbits

### Add support to other microcontrollers

1. Add the a new makefile definitions in "devices" folder, eg, **_myDevice.mk_**
2. Edit the device name, frequency, fuses, lock bits, flash and bootloader parameters.

## Bootloader Logic

### Enter Bootloader Mode
1. The bootloader waits for **_Sync Command_ (SYN [0x16])**. If sync command not received for a while, jump to user program.
2. When receive sync, the bootloader respose with **_Acknowledge Command (ACK [0x06])_** (1 byte), the **_BOOTLOADER VERSION_** (3 bytes), the **_CPU MODEL_** (20 bytes), the **_PAGE_SIZE_** (2 bytes) and the available **_MEMORY_SIZE_** for program (2 bytes), 28 bytes in total (see figure 1).
3. The bootloader wait for the **_Enter Bootloader command (SI [0x0F])_** or the **_Cancel Bootloader command (ESC [0x1B])_**. If some command is not received for a while, jump to user program.
4. If the bootloader receives the **_Cancel Bootloader command_**, jump to user program, else if receives the **_Enter Bootloader command_** enter in bootloader mode
5. The bootloader response with **_Acknowledge Command (ACK [0x06])_** to the **_Enter Bootloader command_** or the **_Cancel Bootloader command_**.
6. Once entered in **_Bootloader mode_**, the bootloader can receive 2 command types: **_Page Write Command (STX [0x02])_** or **_End of Transmission Program (EOT [0x04])_**.

##### Figure 1

	     ACK      VERSION    CPU MODEL   PAGE SIZE   FLASH SIZE
	┌──────────┬───────────┬───────────┬───────────┬────────────┐
	│ 1 (0x06) │     3     │    20     │      2    │      2     | = 28 bytes
	└──────────┴───────────┴───────────┴───────────┴────────────┘

#

### Send a program to bootloader
The loading of a program is done after entering the bootloader mode, sending the **_Page Write Command (STX [0x02])_** followed by the page data chunks of size **_PAGE_SIZE_**. At the end of each packet (STX + DATA_PAGE) the booloader responds with **_Acknowledge Command (ACK [0x06])_**.

The program to be loaded must be fragmented into **_N_** pages of size **_PAGE_SIZE_** data. If the size of the program is not a multiple of **_PAGE_SIZE_**, the remaining data of the program must be added to the last page and filled with **_0xFF_** until the size of **_PAGE_SIZE_** is completed.

The bootloader response with **_No Acknowledge Command (NACK [0x15])_** if **_PAGE_SIZE_** bytes not received in a while after a **_Page Wirte Command (STX [0x02])_**.

Once the loading of the program has been completed, the **_End of Transmission Program (EOT [0x04])_** must be sent to tell the bootloader that the loading has finished. The bootloader responds with **_Acknowledge Command (ACK [0x06])_**  to indicate that it has left the bootloader mode and will execute the program.

If the bootloader receives an invalid command, it will respond with the **_Cancelation Command (CAN [0x18])_**].

1. Send **_Page Write Command (STX [0x02])_**
2. Once the **_Page Write Command_** is received, bootloader responds with **_Acknowledge Command (ACK [0x06])_**
3. Send data in chunks of size **_PAGE_SIZE_**
4. Once the **_PAGE_SIZE_** data is received, bootloader responds with **_Acknowledge Command (ACK [0x06])_**. If the bootloader not receive **_PAGE_SIZE_** chunk in a while, it will respose with **_No Acknowledge Command (NACK [0x15])_** and exit from bootloader mode.
5. Repeat steps 1 to 4 while there is still data to be sent.
6. Send the **_End of Transmission Program (EOT [0x04])_** command to finish loading.
7. The bootloader responds with **_Acknowledge Command (ACK [0x06])_** to indicate that it has left the bootloader mode, and it will jump to the user program.

---

CHANGELOG.md
# ABOS Changelog

## [ 1.2.0 ] - 2018-11-10
- Add mcu model in device info.
- Change de order of the info params to send to the host.
- Add small, normal, medium and big hex files to test with abos-loader.

## [ 1.0.0 ] - 2018-11-10
- Add first version of ABOS supporting Atmega16, Atmega32 and Atmega128 microcontrollers


abos.c
 /*****************************************************************************
 *   file: abos.c                                                             *
 *                                                                            *
 *	 ABOS Bootloader                                                          *
 *                                                                            *
 *   This program is free software: you can redistribute it and/or modify     *
 *   it under the terms of the GNU General Public License as published by     *
 *   the Free Software Foundation, either version 3 of the License, or        *
 *   (at your option) any later version.                                      *
 *                                                                            *
 *   This program is distributed in the hope that it will be useful,          *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of           *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
 *   GNU General Public License for more details.                             *
 *                                                                            *
 *   You should have received a copy of the GNU General Public License        *
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>     *
 *                                                                            *
 *   Written by Alfredo Orozco <alfredoopa@gmail.com>                         *
 *****************************************************************************/

#define BAUD 38400

#include <avr/io.h>
#include <util/delay.h> 
#include <avr/boot.h>
#include <avr/interrupt.h>
#include <avr/pgmspace.h>
#include <util/setbaud.h>
#include <string.h>

#define ABOS_VERSION_MA 1
#define ABOS_VERSION_MI 2
#define ABOS_VERSION_IS 0

#ifdef UDR

#define UC_UDR    UDR
#define UC_UCSRC  UCSRC
#define UC_UCSRB  UCSRB
#define UC_UCSRA  UCSRA
#define UC_UDRE   UDRE
#define UC_RXC    RXC
#define UC_UBRRL  UBRRL
#define UC_UBRRH  UBRRH
#define UC_URSEL  URSEL
#define UC_UCSZ0  UCSZ0
#define UC_UCSZ1  UCSZ1
#define UC_RXEN   RXEN
#define UC_TXEN   TXEN

#elif defined(UDR0)			

#define UC_UDR    UDR0
#define UC_UCSRC  UCSR0C
#define UC_UCSRB  UCSR0B
#define UC_UCSRA  UCSR0A
#define UC_UDRE   UDRE0
#define UC_RXC    RXC0
#define UC_UBRRL  UBRR0L
#define UC_UBRRH  UBRR0H
#define UC_URSEL  0
#define UC_UCSZ0  UCSZ00
#define UC_UCSZ1  UCSZ01
#define UC_RXEN   RXEN0
#define UC_TXEN   TXEN0

#elif defined(UDR1)			

#define UC_UDR    UDR1
#define UC_UCSRC  UCSR1C
#define UC_UCSRB  UCSR1B
#define UC_UCSRA  UCSR1A
#define UC_UDRE   UDRE1
#define UC_RXC    RXC1
#define UC_UBRRL  UBRR1L
#define UC_UBRRH  UBRR1H
#define UC_URSEL  0
#define UC_UCSZ0  UCSZ10
#define UC_UCSZ1  UCSZ11
#define UC_RXEN   RXEN1
#define UC_TXEN   TXEN1

#endif

#define SYNC_COMMAND                0x16
#define ENTER_BOOTLOADER_COMMAND    0x0F
#define END_TRANSMISSION_COMMAND    0x04
#define ACK_COMMAND                 0x06
#define NACK_COMMAND                0x15
#define CANCEL_BOOTLOADER_COMMAND   0x1B
#define PAGE_WRITE_COMMAND          0x02
#define CANCELATION_COMMAND         0x18

#define MAX_RECEIVE_TRIES 700

#define MCU_MODEL_LEN 20

void UsartWrite(uint8_t byte);
int UsartRead();
void Bootloader();

int main(void) {

	char mcu_model[MCU_MODEL_LEN+1] = MCU_MODEL;

	int recv;
	/* Inicializa la USART */
	UC_UBRRL = UBRRL_VALUE;
	UC_UBRRH = UBRRH_VALUE;
	UC_UCSRC = (1 << UC_URSEL) | (1 << UC_UCSZ0) | (1 << UC_UCSZ1);
	UC_UCSRB = (1 << UC_RXEN) | (1 << UC_TXEN);
	
	/* Si se recibe el caracter SYN entra en modo bootloader */
	if(UsartRead() == SYNC_COMMAND)
	{
		/* Envía ACK */
		UsartWrite(ACK_COMMAND);
		/* Envia la versión del ABOS */
		UsartWrite((uint8_t)ABOS_VERSION_MA);
		UsartWrite((uint8_t)ABOS_VERSION_MI);
		UsartWrite((uint8_t)ABOS_VERSION_IS);

		/* Envía el modelo del microcontrolador */
		for(uint8_t i = 0; i < MCU_MODEL_LEN; i++)
			UsartWrite(mcu_model[i]);

		/* Envia el tamaño de página */
		UsartWrite((uint8_t)SPM_PAGESIZE);
		UsartWrite(SPM_PAGESIZE >> 8);

		/* Envia el espacio disponible para el programa */
		UsartWrite((uint8_t)((FLASH_SIZE-BOOT_SIZE)/1024));
		UsartWrite(((FLASH_SIZE - BOOT_SIZE)/1024) >> 8);


		/* Espera comando SYNC para entrar en modo bootloader o 
		   ESC para cancelar la carga */
		while((recv = UsartRead()) != ENTER_BOOTLOADER_COMMAND && recv != CANCEL_BOOTLOADER_COMMAND);
		if(recv == ENTER_BOOTLOADER_COMMAND || recv == CANCEL_BOOTLOADER_COMMAND)
			UC_UDR = ACK_COMMAND;

		/* Entra en modo bootloader */
		if(recv == ENTER_BOOTLOADER_COMMAND)
			Bootloader();	
		/* Espera 5 ms despues de salir del bootloader */
	}

	/* Salta a la aplicación en la dirección 0 */
	asm("JMP 0");
}

void Bootloader()
{
	int cmd;
	uint8_t exit=0;
	uint16_t i;
	uint8_t sreg;
	uint32_t addr=0;
	uint32_t page = 0;

	while(!exit)
	{
		cmd = UsartRead();
		switch(cmd)
		{
			/* Comando STX, inicia la recepción de una página de datos */
			case PAGE_WRITE_COMMAND:
				/* Calcula la dirección dependiendo de la pagina */
				addr = (page++) * SPM_PAGESIZE;
				/* Guarda el registro SREG */
				sreg = SREG;
				/* Deshabilita interrupciones */
				cli();
				/* Espera a que la memoria EEPROM no este ocupada */
				eeprom_busy_wait();
				/* Borra la página de memoria */
				boot_page_erase(addr);
				/* Espera mientras la página de memoria es borrada */
				boot_spm_busy_wait();
				/* Envía ACK para avisar que está listo para recibir datos */
				UsartWrite(ACK_COMMAND);

				/* Recibe la página a programa desde la USART */
				for (i=0; i<SPM_PAGESIZE; i+=2)
				{
					/* Guarda los datos en formato Little Endian */
					uint16_t w1 = UsartRead();
					uint16_t w2 = UsartRead();
					/* Si no se recibieron datos (timeout) envia NACK y sale del modo bootloader */
					if(w1 == -1 || w2 == -1)
					{
						/* Envía NACK */
						UsartWrite(NACK_COMMAND);
						/* Sale del modo bootloader */
						exit = 1; 
						break;
					}
					/* Une los 2 BYTES en un WORD */
					w1 += (w2) << 8;
					/* Llena la página con el dato y la dirección */
					boot_page_fill(addr + i, w1);
				}

				/* Si hubo un error, sale del modo booloader */
				if(exit == 1)
					break;

				/* Escribe la página en la memoria FLash */
				boot_page_write (addr);     
				/* Espera mientras la memoria Flash es escrita */
				boot_spm_busy_wait();       
				/* Habilita la sección RWW. Es necesario si se quiere saltar
				 a la aplicación despues de salir del bootloader */
				boot_rww_enable ();
				// Habilita interrupciones
				sei();
				/* Restaura el registro SREG */
				SREG = sreg;
				/* Envía ACK para indicar que la página fue escrita y 
				puede recibir mas datos */
				UsartWrite(ACK_COMMAND);
			break;

			/* Comando EOT. Finaliza la carga de la aplicación */
			case END_TRANSMISSION_COMMAND:
				/* Envía ACK para indicar que se ha finalizado correctamente */
				UsartWrite(ACK_COMMAND);
				/* Sale del modo bootloader */
				exit = 1;
			break;
			
			/* Si no se recibio un comando válido sale del modo bootloader */
			default: 
				/* Envía CAN, esperando que el cargador lo reciba. */
				UsartWrite(CANCELATION_COMMAND);
				exit=1; 
			break;
		}
	}
}

/* 
	UsartWrite
		Envía un dato por la USART. 

	Recibe:
		Dato de 8 bits a enviar.
*/
void UsartWrite(uint8_t byte){
	/* Si el transmisor está vacío envía el dato */
	while (!(UC_UCSRA & _BV(UC_UDRE)));
		UC_UDR = byte;
}

/* 
	UsartRead
		Lee un dato recibido en la USART.

	Devuelve:
		El dato recibido por la USAR.
		Si no se recibe un dato en un tiempo TIMEOUT devuelve -1.
*/
int UsartRead(){
	uint32_t timeout = 0;
	/* Mientras no se reciba un dato o mientras no se 
	exeda el tiempo de espera, sigue esperando un dato. */
	while (!(UC_UCSRA & _BV(UC_RXC)) && timeout < MAX_RECEIVE_TRIES)
	{
		_delay_us(500);
		timeout++;
	}

	/* Si salio del ciclo por haber excedio el tiempo 
	de espera devuelve -1, si no devuelve el dato recibido. */
	return timeout == MAX_RECEIVE_TRIES ? -1 : UC_UDR;
}
